Feature: Add support for @DatabaseView annotations

mqus opened the pull request:
This adds support for @DatabaseView annotations known from room. One test was added, further testing is needed. Implements #130. I tried to do minimal adjustments to the codebase while copying the handling of entities. I took over the Field handling from there and then just ignored the nullable property by implementing the createStatement function differently. 

As I already said in #259, I may have no idea what I'm doing, so please tell me if I did something wrong or if I can do something better.

I'll try to add some more tests before this is ready to merge. Are there some specific things you would like to see tested?

mqus commented:
The test that's currently failing is the same as in #259 and as far as I can tell, this fails even on the current developer branch. I assume that some underlying framework (dart?) got changed but I couldn't find the issue with my very limited knowledge of `Stream`s.

vitusortner commented:
At a first glance, this looks great 🎉 Thanks for your contribution! I'll review the full PR as soon as possible.

mqus commented:
Some remarks about hacks I did and shouldn't change back myself right now:
- floor_generator relies on floor_annotation via `path: ../floor_annotation` because it needs the new annotations. This has to be changed before the next release but can't be changed right now because nothing would work anymore.
- the same for floor, which relies on floor_generator and floor_annotation.

I also messed up my branches a bit when starting out so view_processor.dart already checks for Uint8List, even if the whole rest of the code can't do anything with it. If you want to merge this PR before #259, I'll remove it. If #259 comes first, I'll rebase and use the proper getter implemented there.

vitusortner commented:
> * floor_generator relies on floor_annotation via `path: ../floor_annotation` because it needs the new annotations. This has to be changed before the next release but can't be changed right now because nothing would work anymore.
> * the same for floor, which relies on floor_generator and floor_annotation.

No problem. Normally, I only change the paths of the library's packages to point to the versions available on pub while creating a release. I must have forgotten to revert it to referring the local version. You can leave it as it is.

> I also messed up my branches a bit when starting out so view_processor.dart already checks for Uint8List, even if the whole rest of the code can't do anything with it. If you want to merge this PR before #259, I'll remove it. If #259 comes first, I'll rebase and use the proper getter there.

Let's merge #259 before.

mqus commented:
I rebased the PR onto the current develop branch and altered `view_processor.dart` as written before. 
I know that an integration test for the Views is still missing and maybe more, do you have any suggestions on what/how should be tested?

mqus commented:
So, I rebased to the current develop and added some more tests (CodeCov seems to work even for PRs now, nice :) )

There is only one open question for me now: There seems to be a mechanism which checks if a table has been altered and then notifies the watchers of the concerning tables. Is that correct? If yes, do you expect this to also work with views? If yes again, then there are three ways to get the tables/entities we have to watch for a specific `DatabaseView`: 

1. We try to detect them in the select query for the view (Seems pretty impossible with all the possibilities to include table names and naming things similarly)
2. Users have to provide the Entities on which a view depends in the `@DatabaseView` annotation. (easy but could lead to an unsafe/leaky way of using the library, since we can't check or guarantee the relation)
3. SQLite/sqflite somehow provides this (I'm not aware of that)
4. Any other Ideas?

Btw, if that watcher mechanism exists, how does it work and how do I access it as a library user?

PS: Ah, right, I should document the DatabaseView usage in the README, or do you want to do it yourself?

mqus commented:
Thank you! I wont disrupt you then ;). Regarding Streams: Afaict Room seems to take the first option and includes a whole sqlite parser...

vitusortner commented on floor_generator/lib/processor/query_method_processor.dart:
```diff
@@ -42,11 +48,16 @@ class QueryMethodProcessor extends Processor<QueryMethod> {
+            (view) =>
+                view.classElement.displayName ==
+                flattenedReturnType.getDisplayString(),
+            orElse: () => null);
```
```suggestion
            orElse: () => null); // doesn't return entity nor view
```

vitusortner commented on floor_generator/lib/processor/query_method_processor.dart:
```diff
@@ -42,11 +48,16 @@ class QueryMethodProcessor extends Processor<QueryMethod> {
-            entity.classElement.displayName ==
-            flattenedReturnType.getDisplayString(),
-        orElse: () => null); // doesn't return an entity
+    final Queryable queryable = _entities.firstWhere(
```
```suggestion
    final queryable = _entities.firstWhere(
```

vitusortner commented on floor_generator/lib/value_object/dao.dart:
```diff
@@ -14,7 +14,7 @@ class Dao {
   final List<DeletionMethod> deletionMethods;
   final List<TransactionMethod> transactionMethods;
-  final List<Entity> streamEntities;
+  final List<Queryable> streamEntities;
```
We only need entities here for now.
```suggestion
  final List<Entity> streamEntities;
```

vitusortner commented on floor_generator/test/processor/view_processor_test.dart:
```diff
@@ -0,0 +1,274 @@
+
+      const expected = "Person(row['id'] as int, row['name'] as String)";
+      expect(actual, equals(expected));
+    });
```
These are basically duplicates from `entity_processor_test`. After migrating duplicate code to `QueryableProcessor`, we could only test this functionality there which would allow us to get rid of this duplication.

mqus replied:
I don't see a simple way to achieve this, tbh, since `QueryableProcessor` is an abstract class. We would have to instantiate the concrete types of `ViewProcessor` and `EntityProcessor` anyway... But if you have a concrete vision on how to do it, please share and I will try.

vitusortner replied:
We could just create an implementation class of `QueryableProcessor` in the test and asserts its methods for correctness.

```dart
class UnderTest extends QueryableProcessor<Object> {
  UnderTest(ClassElement classElement) : super(classElement);

  @override
  Object process() {
    return null;
  }
}
```

mqus replied:
I'll try!

mqus commented on floor_generator/lib/processor/view_processor.dart:
```diff
@@ -0,0 +1,126 @@
+import 'package:floor_generator/value_object/field.dart';
+import 'package:floor_generator/value_object/view.dart';
+
+class ViewProcessor extends Processor<View> {
```
I didn't put `_getName()` in the `QueryableProcessor` because it is pretty dependent on the Type (`View` vs `Entity`). But for the rest that's a pretty good idea.

mqus commented on floor_generator/pubspec.yaml:
```diff
@@ -18,7 +18,7 @@ dependencies:
   collection: ^1.14.11
   floor_annotation:
-    path: ../floor_annotation/
+    path: ../floor_annotation
```
Does that have a functional difference? Or is this for consistency?

vitusortner replied:
Just for consistency 👍

mqus commented on floor/test/integration/database_test.dart:
```diff
@@ -295,6 +299,43 @@ void main() {
         });
       });
+
+      group('Query View', () {
```
So I'll move the NameDao, the Name class and copy all other DAOs and entities to a file in `view_test/` and put the tests there, too (like in blob_test)? This seems pretty redundant to me...

I have the following counter proposal: I'll move just the tests (and maybe the database class) into a separate file (or directory) but will still reference the common database objects.
What do you think?

vitusortner replied:
Sounds great!

mqus commented:
I rebased, integrated most of the suggestions and added an exception if a library user will try to use a Dao method which will stream the results of a query on a database view.

vitusortner commented on floor/test/integration/view_test/view_test.dart:
```diff
@@ -0,0 +1,95 @@
+
+part 'view_test.g.dart';
+
+@Database(version: 2, entities: [Person, Dog], views: [Name])
```
💡We can get of all the migration related code in this file as we're not testing migrations.

```suggestion
@Database(version: 1, entities: [Person, Dog], views: [Name])
```

vitusortner commented on floor/test/integration/view_test/view_test.dart:
```diff
@@ -0,0 +1,95 @@
+      database = await $FloorViewTestDatabase
+          .inMemoryDatabaseBuilder()
+          .addMigrations(allMigrations)
+          .build();
```
```suggestion
      database = await $FloorViewTestDatabase.inMemoryDatabaseBuilder().build();
```

vitusortner commented on floor_generator/lib/processor/queryable_processor.dart:
```diff
@@ -0,0 +1,87 @@
+import 'package:floor_generator/value_object/field.dart';
+import 'package:meta/meta.dart';
+
+abstract class QueryableProcessor<T> extends Processor<T> {
```
💡We could give the generic type some boundaries.

```suggestion
abstract class QueryableProcessor<T extends Queryable> extends Processor<T> {
```

vitusortner commented on floor_generator/lib/value_object/database.dart:
```diff
@@ -26,6 +29,7 @@ class Database {
           name == other.name &&
           entities == other.entities &&
+          views == other.views &&
           daoGetters == other.daoGetters &&
```
💡We should start using `ListEquality` for comparing lists in the database class as well.

```suggestion
          const ListEquality<Entity>().equals(entities, other.entities) &&
          const ListEquality<View>().equals(views, other.views) &&
          const ListEquality<DaoGetter>()
              .equals(daoGetters, other.daoGetters) &&
```

vitusortner commented on floor_generator/lib/value_object/dao.dart:
```diff
@@ -1,11 +1,12 @@
 import 'package:floor_generator/value_object/transaction_method.dart';
 import 'package:floor_generator/value_object/update_method.dart';
 
+import 'entity.dart';
```
💡We could get rid of all the unnecessary import qualifiers.

```dart
import 'deletion_method.dart';
import 'entity.dart';
import 'insertion_method.dart';
import 'query_method.dart';
import 'transaction_method.dart';
import 'update_method.dart';
```

mqus replied:
Then we should do it everywhere... but I think this should be part of another PR... I'll just put the entity import back where it was for now to make int consistent with the other statements...

mqus commented:
I adopted the suggestions from the review and added some documentation to the README files.

tulioccalazans commented:
@vitusortner  When this  feature will be available as a release?

vitusortner commented:
We've just created a new release with the version 0.12.0, which makes database views available and includes other features, fixes and improvements.
